### PR TITLE
[cinder-csi-plugin] Refactor Metadata

### LINKED ISSUE #1030: [cinder-csi-plugin] Availability zone can not be resolved with configdrive
**What happened**:
Deployments with metadata drive will fail to resolve availability zone.

Cinder CSI still uses its own metadata implementation that only works with metadata service and not with configdrive.

**What you expected to happen**:
VMs with configdrive only would be able to get their metadata.

**How to reproduce it**:
VM has only configdrive and not metadata service on network.

**Anything else we need to know?**:

package "k8s.io/cloud-provider-openstack/pkg/util/metadata" implements correct and functional metadata service. So there is no need for CSI own metadata implementation. Whole openstack_metadata.go could be removed.

When metadata is needed it could be fetched with:
`md, err := metadata.Get(ns.Cloud.GetMetadataOpts().SearchOrder)`

Also the default search order should be ConfigDriveID and then MetadataID. As detecting missing device is fast but missing metadata service will take until request timeouts.

**Environment**:
- openstack-cloud-controller-manager(or other related binary) version: master
- OpenStack version: -
- Others: -

/kind bug

### PR DESCRIPTION
**Pull request by ramineni**
**What this PR does / why we need it**:
This PR is to reuse existing metadata interface and remove the redundancy
It also updates the following
1. Removes the use of cloud-init to retrieve instance-id , instead it uses configDrive or MetadataService
2. Support retrieval of AZ from either config drive or metadata service 
3. cleaned up redundant code

**Which issue this PR fixes(if applicable)**:
fixes #1030 

**Special notes for reviewers**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/util/metadata/metadata.go [hunk 1]
```diff
@@ -251,3 +272,21 @@ func Get(order string) (*Metadata, error) {
 	}
 	return metadataCache, nil
 }
+
+// GetInstanceID from metadata service
```
Comment (jichenjc) [inline pkg/util/metadata/metadata.go:276]: the comment is incorrect

File: pkg/util/metadata/metadata.go [hunk 2]
```diff
@@ -251,3 +272,21 @@ func Get(order string) (*Metadata, error) {
...
+func (m *metadataService) GetInstanceID() (string, error) {
+	md, err := Get(m.searchOrder)
+	if err != nil {
+		return "", err
+	}
+	return md.UUID, nil
+}
+
+// GetAvailabilityZone returns zone from metadata service
```
Comment (jichenjc) [inline pkg/util/metadata/metadata.go:285]: ditto

File: pkg/util/metadata/metadata.go [hunk 3]
```diff
@@ -77,6 +82,33 @@ type Metadata struct {
...
+}
+
+// IMetadata implements GetInstanceID & GetAvailabilityZone
+type IMetadata interface {
+	GetInstanceID() (string, error)
+	GetAvailabilityZone() (string, error)
+}
+
+// GetMetadataProvider retrieves instance of IMetadata
+func GetMetadataProvider(order string) (IMetadata, error) {
```
Comment (jichenjc) [inline pkg/util/metadata/metadata.go:96]: looks like this error is useless now

File: pkg/util/metadata/metadata.go (hunk 1 above)
Comment (ramineni) [reply pkg/util/metadata/metadata.go]: done

File: pkg/util/metadata/metadata.go (hunk 3 above)
Comment (ramineni) [reply pkg/util/metadata/metadata.go]: done

**Comment by lingxiankong** [comment]
```
2020-06-16 11:59:03.057720 | ubuntu-bionic | + /home/zuul/src/k8s.io/kubernetes/_output/bin/kubectl --kubeconfig /var/run/kubernetes/admin.kubeconfig get nodes
2020-06-16 11:59:03.146695 | ubuntu-bionic | No resources found
2020-06-16 11:59:03.149875 | ubuntu-bionic | + sleep 2
2020-06-16 11:59:05.152478 | ubuntu-bionic | + wait_time=0
2020-06-16 11:59:05.152624 | ubuntu-bionic | + '[' 0 -gt 0 ']'
2020-06-16 11:59:05.152674 | ubuntu-bionic | + return 1
```

Is this failure in cloud-provider-openstack-acceptance-test-keystone-authentication-authorization also expected?

**Comment by ramineni** [comment]
> ```
> 2020-06-16 11:59:03.057720 | ubuntu-bionic | + /home/zuul/src/k8s.io/kubernetes/_output/bin/kubectl --kubeconfig /var/run/kubernetes/admin.kubeconfig get nodes
> 2020-06-16 11:59:03.146695 | ubuntu-bionic | No resources found
> 2020-06-16 11:59:03.149875 | ubuntu-bionic | + sleep 2
> 2020-06-16 11:59:05.152478 | ubuntu-bionic | + wait_time=0
> 2020-06-16 11:59:05.152624 | ubuntu-bionic | + '[' 0 -gt 0 ']'
> 2020-06-16 11:59:05.152674 | ubuntu-bionic | + return 1
> ```
> 
> Is this failure in cloud-provider-openstack-acceptance-test-keystone-authentication-authorization also expected?

@lingxiankong hmm, webhook service doesn't seem to start, 
0622 05:50:39.386556    9052 webhook.go:199] Failed to make webhook authorizer request: Post https://localhost:8443/webhook?timeout=30s: dial tcp 127.0.0.1:8443: connect: connection refused

I'm not very sure where exactly is failing though , looking it up

**Comment by lingxiankong** [comment]
/override openlab/cloud-provider-openstack-acceptance-test-e2e-conformance
/override openlab/cloud-provider-openstack-acceptance-test-keystone-authentication-authorization

**Comment by lingxiankong** [comment]
Override test-keystone-authentication-authorization is safe, but skipping test-e2e-conformance is a little bit risky, hope we could solve that CI job asap.

/lgtm
/approve

**Pull request merged**